src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs

```diff
@@ -0,0 +1,561 @@
+        // thread token
+        //void ComputeEndpointIdentity()
+        //{
+        //    EndpointIdentity result = null;
+        //    if (this.State == CommunicationState.Opened)
+        //    {
+        //        if (this.SecurityProtocolFactory != null)
+        //        {
+        //            result = this.SecurityProtocolFactory.GetIdentityOfSelf();
+        //        }
+        //        else if (this.SessionServerSettings != null && this.SessionServerSettings.SessionProtocolFactory != null)
+        //        {
+        //            result = this.SessionServerSettings.SessionProtocolFactory.GetIdentityOfSelf();
+        //        }
+        //    }
+        //    if (result == null)
+        //    {
+        //        result = base.GetProperty<EndpointIdentity>();
+        //    }
+        //    this.identity = result;
+        //}
+
+        private void ThrowIfProtocolFactoryNotSet()
+        {
+            if (this.securityProtocolFactory == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.SecurityProtocolFactoryShouldBeSetBeforeThisOperation)));
+            }
+        }
+
+        public async Task<IServiceChannelDispatcher> CreateServiceChannelDispatcherAsync(IChannel outerChannel)
+        { 
+            //TODO should have better logic
+            //Initialization path start
+            if (outerChannel.ChannelDispatcher == null)
+            {
+                TypedChannelDemuxer typedChannelDemuxer = this.ChannelBuilder.ChannelDemuxer.GetTypedServiceDispatcher<IReplyChannel>();
+                IServiceChannelDispatcher channelDispatcher = await typedChannelDemuxer.CreateServiceChannelDispatcherAsync(outerChannel);
+                return channelDispatcher;
+            }
+            //initialization end
+            //Below dispatches all SCT call for first time for all clients
+            else
+            {
+                IServiceChannelDispatcher securityReplyChannelDispatcher = GetInnerChannelDispatcher(outerChannel);
+                return securityReplyChannelDispatcher;
+            }
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="outerChannel"></param>
+        /// <returns></returns>
+        internal IServiceChannelDispatcher GetAuthChannelDispatcher(IReplyChannel outerChannel)
+        {
+            if (securityAuthServiceChannelDispatcher == null)
+            {
+                lock (ThisLock)
+                {
+                    if (securityAuthServiceChannelDispatcher == null)
+                    {
+                        Task<IServiceChannelDispatcher> channelTask = SecurityAuthServiceDispatcher.CreateServiceChannelDispatcherAsync(outerChannel);
+                        securityAuthServiceChannelDispatcher = channelTask.GetAwaiter().GetResult();
+                    }
```

Comment:
Need to make async. You can do this by making the channelTask be an instance field and inside the lock check if channelTask is null and if it is call the async method to populate it. Then outside the lock, await the channelTask which might have been created by a different thread or this thread, it doesn't matter. Then assign the result of awaiting the channelTask to securityAuthServiceChannelDispatcher. You will need to call Thread.MemoryBarrier() after the assignment as you don't have a lock to provide an implicit memory barrier. You will end up with a race to write to securityAuthServiceChannelDispatcher but that's okay as the assignment will be of the same result from the same channelTask instance so it will just be doing the same thing twice.